File: 126-word-ladder-ii/126-word-ladder-ii.py
class Solution:
    def findLadders(self, beginWord: str, endWord: str, wordList: List[str]) -> List[List[str]]:
        
        wordList = set(wordList)
        
        def canBeNext(c, n):
            diffCount = 0
            for i in range(len(c)):
                if c[i] != n[i]:
                    diffCount += 1
            return diffCount == 1
        
        q = [beginWord]
        dep = {beginWord: 0}
        
        while len(q) != 0:
            u = q.pop(0)
            if u == endWord:
                break
            for i in range(len(u)):
                for j in 'qwertyuiopasdfghjklzxcvbnm':
                    v = u[:i] + j + u[i+1:]
                    if v in wordList and v not in dep:
                        dep[v] = dep[u] + 1
                        q.append(v)

        d = 0 if endWord not in dep else dep[endWord]

        if d == 0:
            return []
        
        visited = set()
        
        @cache
        def dfs(u = beginWord, dep = d):
            if dep == 0 and u == endWord:
                return [[u]]
            elif dep > 0:
                ret = []
                for i in range(len(u)):
                    for j in 'qwertyuiopasdfghjklzxcvbnm':
                        v = u[:i] + j + u[i+1:]
                        if v in wordList:
                            ret += dfs(v, dep - 1)
                ret = [[u] + k for k in ret]
                return ret
            else:
                return []
        
        return dfs()
                    

File: 126-word-ladder-ii/126-word-ladder-ii.java
class Solution {
    
    Set<String> wordSet;
    String endWord;
    Map<String, List<List<String>>> cache;
    
    boolean canBeNext(String cur, String next) {
        int diffCount = 0;
        for (int i = 0; i< cur.length(); i++) {
            if (cur.charAt(i) != next.charAt(i))
                diffCount++;
        }
        return diffCount == 1;
    }
    
    
    List<List<String>> dfs(String u, int dep) {
        if (cache.containsKey(u+dep)) {
            return cache.get(u+dep);
        }
        else if (dep == 1 && u.equals(endWord)) {
            List<List<String>> ret = new ArrayList<>();
            List<String> path = new ArrayList<>();
            path.add(u);
            ret.add(path);
            return ret;
        }
        else if (dep == 1) {
            return new ArrayList<>();
        }
        else {
            List<List<String>> all = new ArrayList<>();
            for (String v: wordSet) {
                if (canBeNext(u, v)) {
                    all.addAll(dfs(v, dep - 1));
                }
            }
            List<List<String>> ret = new ArrayList<>();
            for (List<String> l: all) {
                List<String> temp = new ArrayList<>(l);
                temp.add(0, u);
                ret.add(temp);
            }
            cache.put(u+dep, ret);
            return ret;
        }
    }
    
    public List<List<String>> findLadders(String beginWord, String endWord, List<String> wordList) {
        wordSet = new HashSet<>(wordList);
        this.endWord = endWord;
        cache = new HashMap<>();
        Queue<String> q = new LinkedList<>();
        Map<String, Integer> dist = new HashMap<>();
        q.add(beginWord);
        dist.put(beginWord, 1);
        
        while (q.size() != 0) {
            String u = q.poll();
            if (u.equals(endWord)) {
                break;
            }
            for (String v: wordList) {
                if (canBeNext(u, v) && !dist.containsKey(v)) {
                    dist.put(v, dist.get(u) + 1);
                    q.add(v);
                }
            }
        }
        
        int d = dist.getOrDefault(endWord, 0);
        if (d == 0) {
            return new ArrayList<>();
        }
        return dfs(beginWord, d);
        
        
    }
}